Add a JIT interpreter

zdevito opened the pull request:
The separate interpreter is used to graphs with a lower overhead than
converting them to autograd graphs. Some notes:

* does not support Handles/PythonOp/CppOp, these will be in a future commit
* jit_closure.cpp still exists and we fall back to it for now when
  cannot handle something because of PythonOp/CppOp
* In order to support retain_graph=True, the interpreter can be cloned,
  creating a copy that can be run with different arguments. This is
  assumed to be the non-standard case so cloning is not particularly optimized.
  No tensor _data_ is copied, but the at::Tensor list in the interpreter is.
  If we hit problems, there is a lot we could do (such as register allocation)
  to minimize the stuff that needs to be copied.
* Uses a pImpl pattern to keep implementation details out of its header file.
* Modifies the way getTensorOp works so that it reads/writes to already-existing
  vectors, this prevents needing to realloc these buffers each time.
* Timings are here: https://gist.github.com/zdevito/5a20ac29fb1b9e449e693b67dc478127
  This reduces overhead to about the same as running it in python.
  It is about 10us faster to run the same thing using ATen directly.

zdevito committed:
Add a JIT interpreter

The separate interpreter is used to graphs with a lower overhead than
converting them to autograd graphs. Some notes:

* does not support Handles/PythonOp/CppOp, these will be in a future commit
* jit_closure.cpp still exists and we fall back to it for now when
  cannot handle something because of PythonOp/CppOp
* In order to support retain_graph=True, the interpreter can be cloned,
  creating a copy that can be run with different arguments. This is
  assumed to be the non-standard case so cloning is not particularly optimized.
  No tensor _data_ is copied, but the at::Tensor list in the interpreter is.
  If we hit problems, there is a lot we could do (such as register allocation)
  to minimize the stuff that needs to be copied.
* Uses a pImpl pattern to keep implementation details out of its header file.
* Modifies the way getTensorOp works so that it reads/writes to already-existing
  vectors, this prevents needing to realloc these buffers each time.
* Timings are here: https://gist.github.com/zdevito/5a20ac29fb1b9e449e693b67dc478127
  This reduces overhead to about the same as running it in python.
  It is about 10us faster to run the same thing using ATen directly.

zdevito commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -204,7 +204,9 @@ static variable_list call_function(FunctionTask& task) {
     if (!callback(&fn, inputs)) return variable_list(fn.next_functions.size());
   }
-
+  if(!task.base->keep_graph) {
```
The interpreter needs to know _before_ it executes that it should be releasing variables. If you don't like this way, I am open to others, but it does need to know ahead of time for efficient execution and small memory footprint.

ezyang replied:
This seems fine to me.

zdevito commented on torch/csrc/jit/interpreter_autograd_function.h:
```diff
@@ -0,0 +1,37 @@
+  virtual void willReleaseVariables() override {
+    keep_graph = false;
+  }
+  virtual autograd::variable_list apply(const autograd::variable_list& inputs) override {
```
I a not confident that Variable flags are being propagated correctly as the interpreter runs. @apaszke can you let me know what I am suppose to be doing with them here. I imagine we should be just saving the flags from the trace and reproducing them here.

ezyang replied:
Actually, assuming that you've accurately got all the inputs, I think this actually will DTRT.

apaszke replied:
No, this won't work. We have the [information necessary to reproduce the flags in `TracingState`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tracer_state.h#L68), so it's possible to do right now. I'm fine with merging as is and I can fix that later.

ezyang commented on torch/csrc/jit/interpreter.h:
```diff
@@ -0,0 +1,50 @@
+
+struct Interpreter {
+  Interpreter(const Function & function);
+  void runOneStage(
```
You can tell this is the case when you look at the 'clone' method below, but there should be comments here saying that an interpreter is a mutable resource which is not thread safe and can only be run once. Perhaps there is another name rather than Interpreter which expresses this more clearly, like "InterpreterHandle"

zdevito replied:
I'll add some comments. I will probably rename Interpreter -> InterpreterState to make it more clear. `lua_State` is a similar object.

ezyang commented on torch/csrc/jit/interpreter.h:
```diff
@@ -0,0 +1,50 @@
+struct InterpreterImpl;
+struct Graph;
+
+struct Function {
```
Blah, this is the same name as autograd Function, which will make it harder to grep for instances of this particular concept. How about `GraphBCO` (graph bytecode object).

ezyang replied:
(Yes, I know we don't actually have a bytecode today. But we might, someday...)

zdevito replied:
I did Function -> Code, and Interpreter -> InterpreterState.

ezyang commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -0,0 +1,288 @@
+  IR_ELSEIF(Constant)
+    auto t = value->t(kvalue);
+    return [t](InputList inputs, OutputList outputs) {
+      outputs.push_back(t);
```
Err... do we allow inplace operations in the new interpreter? If so, wouldn't we get in trouble if we did an inplace op on this constant? I remember thinking we had this hazard in the old code too, but we had a clone to prevent this from happening:

```
struct WrapConstant : public Function {
  WrapConstant(at::Tensor value)
    : value(std::move(value)) {
    is_executable = true;
    num_inputs = 1;
  }

  virtual variable_list apply(const variable_list& inputs) {
    if (inputs.size() != 1 || inputs[0].defined())
      throw std::logic_error("WrapConstant nodes should only receive a single NULL input");
    AutoGPU guard(value);
    return {make_variable(value.clone())};
  }

  at::Tensor value;
};
```

zdevito replied:
I am assuming the Graph doesn't have any inplace operators. What we can do later is optimize the instruction generation to re-use inputs tensors when an inplace variant exists and it is safe to do (no more uses, not ruining a constant). Copying constants can potentially add huge overhead, so it makes sense to stick with working with the fast path.

ezyang commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -0,0 +1,288 @@
+// contiguous we allocate a single vector and use offsets into the vector
+// which are stored in the RegList struct
+// start is an offset into int_data of Function if this list is integers
+// and bool_data of Function if this list is booleans (only free_flags)
```
I'm annoyed that there is no static or dynamic information telling me whether or not a RegList struct indexes into int_data or bool_data. Can we at least get comments saying which array we're indexing into?

ezyang replied:
(The overengineer in me wonders if you couldn't use a custom allocator to get everything contiguous... probably not worth it haha)

ezyang replied:
After reading more code, I also wonder if this isn't misnamed. RegList struct is being used to universally represent "array in contiguous sequence of memory", it really isn't a "list of registers" at all.

zdevito replied:
Yeah, we can probably add some more infra to make this clearer.

ezyang commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -0,0 +1,288 @@
+
+struct UseList {
+  // values to be used
+  RegList values;
```
What is a "value"? (doc request)

ezyang commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -0,0 +1,288 @@
+          break;
+        // unused inputs are given a false register -1 so that we never hold a
+        // reference to the tensor data, otherwise we would fail to clean them
+        // up since they do not have a last use at which to free them
```
Nice.

ezyang commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -0,0 +1,288 @@
+#include "torch/csrc/jit/fusion_compiler.h"
+#endif
+
+namespace torch { namespace jit {
```
It'd be positively lovely if you could record down which perf optimizations you *know* pushed the needle down, if you remember this from when you were working on the interpreter. (Unless, well, it sprang fully formed from your head as fast as it is today. That's cool too 8)

zdevito replied:
I didn't add optimizations after the fact. It's easier to start with good perf than try to optimize later. Instead there were two goals:

1. Don't allocate memory when you running (there is a single malloc when the Interpreter is created)
2. Scan linearly through memory as much as possible to be friendly to cache.

It is not clear if either of these were necessary but in some sense it is easier writing it this way, then writing it another way and then trying all optimizations.

ezyang commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -85,16 +85,16 @@ def is_tensor_arg(arg):
             if any(arg['simple_type'] == 'TensorList' for arg in arguments):
                 assert sum(map(is_tensor_arg, arguments)) == 1
-                args = ['as_tensor_list(vars)' if is_tensor_arg(arg) else arg['name']
+                args = ['inputs' if is_tensor_arg(arg) else arg['name']
```
No one expects the Spanish Implicit Conversion!!

zdevito replied:
I changed all of these methods to actually work Tensor -> Tensor rather than Variable -> Variable. So this is no longer a cast.

ezyang reviewed (approved):
This is so groovy! As usual, you should make the final call whether or not you want to address the comments I made or not; the only outright potential bug is us not cloning tensor constants from the graph IR when we construct an interpreter state.

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -0,0 +1,288 @@
+      for(auto output : node->outputs()) {
+        intListInsert(inst.outputs, getOrAllocateRegister(output));
+      }
+      inst.callback = getCallback(node);
```
Wouldn't it be more natural to just do
```cpp
Instruction inst;
// initialize everything here
stages.back().instructions.push_back(std::move(inst));
```

apaszke commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -0,0 +1,288 @@
+      }
+      inst.callback = getCallback(node);
+    }
+    insertStagesTo(cur_stage, graph->stage(), input_pos, output_pos);
```
Why is this necessary?

zdevito replied:
It is possible (though unlikely), that a final stage is just an identity function of its inputs so there won't be instructions. So this makes sure we have inserted Stage objects for each stage.

apaszke replied:
Can you please add a comment? It's not immediately obvious

apaszke reviewed (approved):
This looks really good. I left some questions. In general I think the code could be simplified a bit (especially `RegList`s are quite unnatural, but I can address that in a next PR).

My biggest concern is the pImpl pattern. Why do we need this? We only have a single implementation, so why add the level of indirection?

apaszke commented on torch/csrc/jit/interpreter_autograd_function.h:
```diff
@@ -0,0 +1,37 @@
+    for(auto & i : inputs) {
+      tinputs.push_back(i.data());
+    }
+    Interpreter interp = (keep_graph) ? interp_.clone() : interp_;
```
This will call the copy constructor of `interp` - probably not what you wanted.

zdevito replied:
interp is just a handle to a `std::shared_ptr<InterpreterImpl>` so copy constructor does not make a full copy. I'll think about whether there is a way to make this more clear.

apaszke replied:
Ah right

zdevito commented on torch/csrc/jit/test_jit.cpp:
```diff
@@ -246,7 +247,204 @@ void internedStringsTests () {
+    std::tie(hx, cx) = lstm(input[0], hx, cx, w_ih, w_hh);
+
+    //std::cout << almostEqual(outputs[0],hx) << "\n";
+    JIT_ASSERT(almostEqual(outputs[0],hx));
```
That's a good point, I wonder if they are. I just always work with non-determinism as my assumed default.

zdevito commented on torch/csrc/jit/test_jit.cpp:
```diff
@@ -246,7 +247,204 @@ void internedStringsTests () {
 
 
+
+at::Tensor t_use(at::Tensor x) {
```
Yeah, I have another patch that just adds the benchmarking code right here as well. :)

zdevito committed:
Code Mod

Interpreter -> InterpreterState
Function -> Code

Add other requested comments.

zdevito committed:
RegList -> ListHandle<T>

Change the RegList functions to be safer by identifying the type of
each argument list, and checking that list insert does not try
to add to two different lists at once.

zdevito committed:
Use exactly equal for interp tests

zdevito commented:
I made the RegList stuff more type safe so you no longer have to guess Bool/Int and you have to try harder to screw up appending to a list.  I tried to do more aggressive changes like create a ContiguousListPool objects, there are some optimizations that remove indirections when accessing these lists that make these changes difficult to encapsulate.
